### DISCUSSION TITLE
Pluggable Database

### DISCUSSION THREAD

**Discussion by machship-mm** [zitadel/zitadel#2708] [Ideas]
Hi guys, 

I'm interested to know a bit more about your choice of using `cockroachdb` as the database. 

Are there unique `cockroachdb` features that you use, or was this choice primarily around its infrastructure concerns (scalability, reliability, etc.)?

The reason I ask is that we use `postgres` and have already got reliable persist database storage handled, so adding `cockroachdb` will add to the management surface of our clusters. I know that `cockroachdb` primarily uses the same syntax and communication protocol as `postgres`, so if you don't use any unique `cockroachdb` features in Zitadel, I might try pointing Zitadel to postgres and see how things run!

Keen to hear about your decision process here :)

**Comment by fforootd** [comment]
Hi @machship-mm 

Great question. I totally see why customers with `postgres` knowledge don't want to invest in running `CockroachDB`.

The primary reason for us to rely on `CockroachDB` are around the ease of operations and especially running on `Kubernetes`. 
What we think is easier to do with  `CockroachDB` :

- Easy to ...
  - rolling restart nodes under load
  - linear scale infra when needed
  - run across availability zones
  - under load version upgrades
  - distribute active zitadel tenants across a cluster
  - distributed backup/restore to gcs/s3 speed

Also, one of the features we took advantage of, is the ["online schema changes"](https://www.cockroachlabs.com/docs/stable/online-schema-changes.html).

There are also some drawbacks:

- Follower reads, data locality and partial index are only available on the payed version 😢 

Call me interest in hearing from you if it works with  `postgres` out of the box 😁 

@adlerhurst did I miss something?

**Reply by adlerhurst** [reply]
I don't think you missed anything. 
As far as i know we don't use any cockroach specific statements. If i have some spare time this week i will try to replace cockroach with postgres.

**Reply by machship-mm** [reply]
Thanks guys!

**Reply by adlerhurst** [reply]
hi @machship-mm 

Sorry for the late reply.

At the moment it's impossible to execute the migrations without any change because we use hash sharded indexes on our `events`-table. 

Enjoy the pre Christmas time.

**Reply by adlerhurst** [reply]
hi @machship-mm 

Finally ZITADEL supports postgresql! It was a bit more of work than we expected then. If you are still interested, I would be happy to get your feedback :-) 

The release notes can be found here: https://github.com/zitadel/zitadel/releases/tag/v2.3.0

**Comment by fforootd** [comment]
Relates to the started effort here #3598
